pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -829,9 +920,147 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
+					klog.Errorf("updateContainerLimits for pod %q resource %s failed: %v", format.Pod(pod), rName, err)
+					return err
+				}
+			}
+			if newPodCgLimit < currPodCgLimit {
+				err = setPodCgroupLimit(rName)
+			}
+			return err
+		}
+		if len(podContainerChanges.ContainersToUpdate[v1.ResourceMemory]) > 0 || len(podContainerChanges.ContainersToRestart) > 0 {
+			currentPodMemoryLimit, err := pcm.GetPodCgroupMemoryLimit(pod)
+			if err != nil {
+				klog.Errorf("GetPodCgroupMemoryLimit for pod %s failed: %v", pod.Name, err)
+				return
+			}
+			if errResize := resizeContainers(v1.ResourceMemory, int64(currentPodMemoryLimit), *podResources.Memory); errResize != nil {
+				return
+			}
+		}
+		if len(podContainerChanges.ContainersToUpdate[v1.ResourceCPU]) > 0 || len(podContainerChanges.ContainersToRestart) > 0 {
+			currentPodCpuQuota, _, _, err := pcm.GetPodCgroupCpuLimit(pod)
+			if err != nil {
+				klog.Errorf("GetPodCgroupCpuLimit for pod %s failed: %v", pod.Name, err)
+				return
+			}
+			if errResize := resizeContainers(v1.ResourceCPU, currentPodCpuQuota, *podResources.CpuQuota); errResize != nil {
+				return
+			}
+		}
+		for _, idx := range podContainerChanges.ContainersToRestart {
+			start("container", containerStartSpec(&pod.Spec.Containers[idx]))
+		}
+	}
+
 	return
 }
 
+func (m *kubeGenericRuntimeManager) updateContainerLimits(pod *v1.Pod, podStatus *kubecontainer.PodStatus, resourceName v1.ResourceName, containersToUpdate []containerToUpdateInfo) error {
+	var updatedLimitsMap map[kubecontainer.ContainerID]struct {
+		cStatus *kubecontainer.ContainerStatus
+		rLimits v1.ResourceList
+	}
+	updatedLimitsMap = make(map[kubecontainer.ContainerID]struct {
+		cStatus *kubecontainer.ContainerStatus
+		rLimits v1.ResourceList
+	})
+
+	for _, cInfo := range containersToUpdate {
+		if *cInfo.restart == true {
+			continue
+		}
+		container := &pod.Spec.Containers[cInfo.specIndex]
+		specLimits := container.Resources.Limits
+		defer func() {
+			container.Resources.Limits = specLimits
+		}()
+		switch resourceName {
+		case v1.ResourceCPU:
+			container.Resources.Limits = v1.ResourceList{
+				v1.ResourceCPU:    *specLimits.Cpu(),
+				v1.ResourceMemory: *cInfo.apiContainerStatus.Resources.Limits.Memory(),
+			}
+		case v1.ResourceMemory:
+			container.Resources.Limits = v1.ResourceList{
+				v1.ResourceCPU:    *cInfo.apiContainerStatus.Resources.Limits.Cpu(),
+				v1.ResourceMemory: *specLimits.Memory(),
+			}
+		}
+		if err := m.updateContainerResources(pod, container, cInfo.kubeContainerStatus.ID); err != nil {
+			klog.Errorf("updateContainerResources %q(id=%q) for pod %q failed: %v", container.Name, cInfo.kubeContainerStatus.ID, format.Pod(pod), err)
+			return err
+		}
+		cInfo.apiContainerStatus.Resources.Limits = container.Resources.Limits
+		// NOTE: Ideally, cpu and memory resource limits come from ContainerStatus CRI API query.
```

Comment:
just to confirm, this means that a failed update of container limits still appears as a success to observers of the pod object?

Reply:
No, if updateContainerResources returns failure at any point, we abort any subsequent updates and exit, and let the next SyncPod attempt retry this at a later time. It's only when all of the planned updates have succeeded that we execute the below GetPodStatus, update PLEG cache code. That update gets propagated to v1 ContainerStatus that user can see.